### DISCUSSION TITLE
How does the seed work?

### DISCUSSION THREAD

**Discussion by Minozzzi** [faker-js/faker#1195] [Q&A] [answered]
Hi guys, I have a question. I want to know how the seed works, when running the tests to generate values.

Thankss :D

**Comment by xDivisionByZerox** [comment, marked as answer]
Well, seeding a `Faker` instance with the same value equals in same return values from different instances (or even the same one).

Example:
```
const one = new Faker(); // for simplicity without actual parameters
const two = new Faker(); // for simplicity without actual parameters

const mySeed = 1337;
one.seed(mySeed);
one.name.firstName(); // Devyn
one.name.firstName(); // Craig

two.seed(mySeed);
two.name.firstName(); // Devyn
two.name.firstName(); // Craig

// reset to get same values "again"
one.seed(mySeed);
one.name.firstName(); // Devyn
one.name.firstName(); // Craig
``` 

[Stackblitz example](https://stackblitz.com/edit/faker-js-demo-pvbyam?file=index.ts)

If you need anything more specific, you might want to be more specific on your question.

**Reply by Minozzzi** [reply]
Yeah, it makes sense. Can the seed number be any number? And, how does the faker.seed function work? How generate the values

**Reply by Minozzzi** [reply]
I say, how faker generates the `faker.firstName() // Devyn`?

**Reply by xDivisionByZerox** [reply]
I have a really hard time following you. We are an open-source project. If you want to see how anything is implemented, you can simply check the source code. In the case of `firstName` you can simply check the address module.

https://github.com/faker-js/faker/blob/ca7cde59ec3747a93d1d47355a83287696628eb5/src/modules/name/index.ts#L84-L93

**Reply by xDivisionByZerox** [reply]
Maybe this will help you. In the example `faker.firstName() // Devyn`, the name `Devyn` is not "generated" from randomly chosen characters, but picked from a (very) long list of names. For English you can find this list of names in the following file:
https://github.com/faker-js/faker/blob/ca7cde59ec3747a93d1d47355a83287696628eb5/src/locales/en/name/first_name.ts#L1-L10

**Reply by Minozzzi** [reply]
I saw how its implemented, but on https://github.com/faker-js/faker/blob/ca7cde59ec3747a93d1d47355a83287696628eb5/src/modules/mersenne/twister.ts#L168 I don't understand.
Does my question make sense?

**Reply by xDivisionByZerox** [reply]
Ok, I see (I hope). The file you are referring to is an implementation of a pseudorandom number generator called [Mersenne Twister](https://en.wikipedia.org/wiki/Mersenne_Twister). It's basically a bunch of math that ensures that if you put numbers in (the seed). The same thing comes out but still looks random. If you want to find out more you should read the article I linked to you or generally any other article about random number generation. 

But TBH I don't think most users don't need to have that deep of an understanding of how this works, since its high-level mathematics and that's kinda scary.

I really hope this helps (sorry if it doesn't :( ).

**Reply by Minozzzi** [reply]
Yes, it really makes sense to me. Thanks for answering me. With this lot of calculation it generates the random values of the locations?
